src/openms/source/ANALYSIS/ID/PrecursorPurity.cpp

```diff
@@ -0,0 +1,231 @@
+      {
+        target_intensity += isolated_window[next_iso_index].getIntensity();
+
+        // std::cout << isolated_window[next_iso_index].getMZ() << " : matched | ";
+
+        isolated_window.erase(isolated_window.begin()+next_iso_index);
+        target_peak_count++;
+      }
+      // always increment iso to progress the loop
+      iso++;
+    }
+    // std::cout << std::endl;
+
+    // std::cout << "noise peaks: ";
+    double noise_intensity(0);
+    for (auto peak : isolated_window)
+    {
+      noise_intensity += peak.getIntensity();
+      // std::cout << peak.getMZ() << " | ";
+    }
+    // std::cout << std::endl;
+
+    double rel_sig(0);
+    if (target_intensity > 0.0)
+    {
+      rel_sig = target_intensity / total_intensity;
+    }
+
+    score.total_intensity = total_intensity;
+    score.target_intensity = target_intensity;
+    score.residual_intensity = noise_intensity;
+    score.signal_proportion = rel_sig;
+    score.target_peak_count = target_peak_count;
+    score.residual_peak_count = isolated_window.size();
+
+    return score;
+  }
+
+  PrecursorPurity::PurityScores PrecursorPurity::combinePrecursorPurities(const PrecursorPurity::PurityScores& score1, const PrecursorPurity::PurityScores& score2)
+  {
+    PrecursorPurity::PurityScores score;
+    score.total_intensity = score1.total_intensity + score2.total_intensity;
+    score.target_intensity = score1.target_intensity + score2.target_intensity;
+    score.residual_intensity = score1.residual_intensity + score2.residual_intensity;
+    if (score.target_intensity > 0.0) // otherwise default value of 0 is used
+    {
+      score.signal_proportion = score.target_intensity / score.total_intensity;
+    }
+    score.target_peak_count = score1.target_peak_count + score2.target_peak_count;
+    score.residual_peak_count = score1.residual_peak_count + score2.residual_peak_count;
+
+    return score;
+  }
+
+  std::vector<PrecursorPurity::PurityScores> PrecursorPurity::computePrecursorPurities(const PeakMap& spectra, double precursor_mass_tolerance, bool precursor_mass_tolerance_unit_ppm)
+  {
+    std::vector<PrecursorPurity::PurityScores> purityscores;
+
+    // TODO throw an exception or at least a warning?
+    // if there is no MS1 before the first MS2, the spectra datastructure is not suitable for this function
+    if (spectra[0].getMSLevel() == 2)
+    {
+      LOG_WARN << "Warning: Input data not suitable for Precursor Purity computation. Will be skipped!\n";
+      return purityscores;
+    }
+
+    // keep the index of the two MS1 spectra flanking the current group of MS2 spectra
+    Size current_parent_index = 0;
+    Size next_parent_index = 0;
+    bool lastMS1(false);
+    for (Size i = 0; i < spectra.size(); ++i)
+    {
+      // change current parent index if a new MS1 spectrum is reached
+      if (spectra[i].getMSLevel() == 1)
+      {
+        current_parent_index = i;
+      }
+      else if (spectra[i].getMSLevel() == 2)
```

Comment:
have you thought about MS3 spectra?

Reply:
@hroest the function to do this for one spectrum is agnostic to the MS level. I guess one possibility would be to add the MS levels as input parameters to the looping function.